File: ios-native/Sprout/Model/Care.swift
import Foundation

/// Подкормка и пересадка: сколько дней сада прошло с последней и раз в
/// сколько их делать. Счёт — в днях сада, как и влажность: сад живёт в
/// ускоренных часах, и настоящие даты разошлись бы с процентами. Подкормка
/// считается только в пору роста — зимой удобрение жжёт корни.
struct Care: Codable, Hashable, Sendable {
    /// Раз в сколько дней подкармливать; пусто — не напоминать.
    var feedEvery: Double?
    var sinceFed: Double = 0

    /// Раз в сколько дней пересаживать; пусто — не напоминать.
    var repotEvery: Double?
    var sinceRepot: Double = 0

    /// Суккуленты едят раз в месяц и живут в одном горшке по два года,
    /// цветущие просят еды каждые две недели, остальные — раз в три.
    static func usual(for preset: Preset) -> Care {
        switch preset {
        case .cactus, .aloe, .echeveria, .jade, .sansevieria, .zamioculcas:
            Care(feedEvery: 30, repotEvery: 730)
        case .orchid:
            Care(feedEvery: 21, repotEvery: 730)
        case .violet, .begonia, .pelargonium, .tulip, .herbs:
            Care(feedEvery: 14, repotEvery: 365)
        default:
            Care(feedEvery: 21, repotEvery: 365)
        }
    }

    /// Дней до подкормки; пусто — не подкармливают.
    var feedIn: Double? { feedEvery.map { max(0, $0 - sinceFed) } }

    var repotIn: Double? { repotEvery.map { max(0, $0 - sinceRepot) } }

    var feedDue: Bool { feedIn.map { $0 < 0.5 } ?? false }

    var repotDue: Bool { repotIn.map { $0 < 0.5 } ?? false }

    mutating func pass(days: Double, growing: Bool) {
        guard days > 0 else { return }
        if growing { sinceFed += days }
        sinceRepot += days
    }

    /// Сроки пересадки на выбор — в месяцах.
    static let repotMonths = [6, 12, 18, 24]

    static func days(months: Int) -> Double { Double(months) * 365 / 12 }

    static func months(days: Double) -> Int {
        Int((days * 12 / 365).rounded())
    }

    /// «Подкормка через 5 дней», «Пора подкормить».
    var feedLabel: String? {
        guard let left = feedIn else { return nil }
        if feedDue { return Lang.text("Пора подкормить") }
        return Lang.format("Подкормка через %lld дней", Int(left.rounded()))
    }

    var repotLabel: String? {
        guard let left = repotIn else { return nil }
        if repotDue { return Lang.text("Пора пересадить") }
        let months = Care.months(days: left)
        guard months >= 1 else {
            return Lang.format("Пересадка через %lld дней",
                               Int(left.rounded()))
        }
        return Lang.format("Пересадка через %lld месяцев", months)
    }
}

File: android/app/src/main/kotlin/com/ledro6/sprout/model/Care.kt
package com.ledro6.sprout.model

import kotlinx.serialization.Serializable
import kotlin.math.max

/**
 * Подкормка и пересадка: сколько дней сада прошло с последней и раз в сколько
 * их делать. Счёт — в днях сада, как влажность. Подкормка считается только в
 * пору роста — зимой удобрение жжёт корни.
 */
@Serializable
data class Care(
    val feedEvery: Double? = null,
    val sinceFed: Double = 0.0,
    val repotEvery: Double? = null,
    val sinceRepot: Double = 0.0,
) {
    val feedIn: Double? get() = feedEvery?.let { max(0.0, it - sinceFed) }

    val repotIn: Double? get() = repotEvery?.let { max(0.0, it - sinceRepot) }

    val feedDue: Boolean get() = feedIn?.let { it < 0.5 } ?: false

    val repotDue: Boolean get() = repotIn?.let { it < 0.5 } ?: false

    fun passed(days: Double, growing: Boolean): Care {
        if (days <= 0) return this
        return copy(
            sinceFed = if (growing) sinceFed + days else sinceFed,
            sinceRepot = sinceRepot + days,
        )
    }

    val feedLabel: String?
        get() {
            val left = feedIn ?: return null
            if (feedDue) return Lang.text("Пора подкормить")
            return Lang.format("Подкормка через %lld дней", left.roundedInt())
        }

    val repotLabel: String?
        get() {
            val left = repotIn ?: return null
            if (repotDue) return Lang.text("Пора пересадить")
            val months = months(left)
            if (months < 1) return Lang.format("Пересадка через %lld дней", left.roundedInt())
            return Lang.format("Пересадка через %lld месяцев", months)
        }

    companion object {
        /** Суккуленты едят раз в месяц, цветущие — раз в две недели. */
        fun usual(preset: Preset): Care = when (preset) {
            Preset.CACTUS, Preset.ALOE, Preset.ECHEVERIA, Preset.JADE, Preset.SANSEVIERIA,
            Preset.ZAMIOCULCAS, Preset.HAWORTHIA, Preset.OPUNTIA, Preset.KALANCHOE, Preset.YUCCA,
            -> Care(feedEvery = 30.0, repotEvery = 730.0)
            Preset.ORCHID, Preset.ANTHURIUM, Preset.HOYA -> Care(feedEvery = 21.0, repotEvery = 730.0)
            Preset.CITRUS -> Care(feedEvery = 14.0, repotEvery = 730.0)
            Preset.VIOLET, Preset.BEGONIA, Preset.PELARGONIUM, Preset.TULIP, Preset.HERBS,
            Preset.ROSE, Preset.MINT, Preset.ROSEMARY, Preset.LILY, Preset.SUNFLOWER,
            Preset.LAVENDER, Preset.CHRYSANTHEMUM,
            -> Care(feedEvery = 14.0, repotEvery = 365.0)
            else -> Care(feedEvery = 21.0, repotEvery = 365.0)
        }

        /** Сроки пересадки на выбор — в месяцах. */
        val repotMonths = listOf(6, 12, 18, 24)

        fun days(months: Int): Double = months * 365.0 / 12

        fun months(days: Double): Int = (days * 12 / 365).roundedInt()
    }
}
